src/task.ts

```diff
@@ -0,0 +1,450 @@
+ * for more information.
+ *
+ * All pending timers must be completed or canceled for an orchestration to
+ * complete.
+ *
+ * @example Cancel a timer
+ * ```javascript
+ * // calculate expiration date
+ * const timeoutTask = context.df.createTimer(expirationDate);
+ *
+ * // do some work
+ *
+ * if (!timeoutTask.isCompleted) {
+ *     timeoutTask.cancel();
+ * }
+ * ```
+ *
+ * @example Create a timeout
+ * ```javascript
+ * const now = Date.now();
+ * const expiration = new Date(now.valueOf()).setMinutes(now.getMinutes() + 30);
+ *
+ * const timeoutTask = context.df.createTimer(expirationDate);
+ * const otherTask = context.df.callActivity("DoWork");
+ *
+ * const winner = yield context.df.Task.any([timeoutTask, otherTask]);
+ *
+ * if (winner === otherTask) {
+ *     // do some more work
+ * }
+ *
+ * if (!timeoutTask.isCompleted) {
+ *     timeoutTask.cancel();
+ * }
+ * ```
+ */
+export interface ITimerTask extends ITask {
+    /**
+     * @returns Whether or not the timer has been canceled.
+     */
+    isCancelled: boolean;
+    /**
+     * Indicates the timer should be canceled. This request will execute on the
+     * next `yield` or `return` statement.
+     */
+    cancel: () => void;
+}
+
+/**
+ * @hidden
+ * Base class for all Tasks, defines the basic state transitions for all tasks.
+ */
+export abstract class TaskBase {
+    public state: TaskState;
+    public parent: CompoundTask | undefined;
+    public isPlayed: boolean;
+    public result: unknown;
+
+    /**
+     * @hidden
+     *
+     * Construct a task.
+     * @param id
+     *  The task's ID
+     * @param action
+     *  The task's backing action
+     */
+    constructor(public id: TaskID, protected action: BackingAction) {
+        this.state = TaskState.Running;
+    }
+
+    /** Get this task's backing action */
+    get actionObj(): BackingAction {
+        return this.action;
+    }
+
+    /** Get this task's current state */
+    get stateObj(): TaskState {
+        return this.state;
+    }
+
+    /** Whether this task is not in the Running state */
+    get hasResult(): boolean {
+        return this.state !== TaskState.Running;
+    }
+
+    get isFaulted(): boolean {
+        return this.state === TaskState.Failed;
+    }
+
+    get isCompleted(): boolean {
+        return this.state === TaskState.Completed;
+    }
+
+    /** Change this task from the Running state to a completed state */
+    private changeState(state: TaskState): void {
+        if (state === TaskState.Running) {
+            throw Error("Cannot change Task to the RUNNING state.");
+        }
+        this.state = state;
+    }
+
+    /** Attempt to set a result for this task, and notifies parents, if any */
+    public setValue(isError: boolean, value: unknown): void {
+        let newState: TaskState;
+
+        if (isError) {
+            if (value instanceof Error) {
+                if (value instanceof TaskBase && value.result instanceof Error) {
```

Comment:
Looks like my comment from a previous iteration about the above two lines got missed in your refactor.  `value` cannot be both an instance of `Error` and an instance of `TaskBase` at the same time.

Reply:
Oh my, I had totally mean to fix this! A lot of things changed in my refactor, including the name of this file, so that must be why I missed this. My next commit includes the correct version (just checking for `instanceof` Error)